src/cursor.jl

```diff
@@ -546,10 +494,8 @@ Return a child cursor vector of the given cursor.
 function children(cursor::CXCursor)
     # TODO: possible to use sizehint! here?
     list = CXCursor[]
-    GC.@preserve cursor begin
```

Comment:
If you are following the normal convention then you need to preserve list but not cursor

Reply:
Just to make sure I understood this correctly. It's necessary to preserve `list` because in the case when the return value of `children` is discarded, the compiler might optimize it out?

Reply:
No. `list` can be freed whenever `pointer_from_objref(list)` returns.

Reply:
Also, this code will be much better written as.

```
function cu_children_visitor(cursor::CXCursor, parent::CXCursor, list)::Cuint
    push!(list, cursor)
    return CXChildVisit_Continue
end

function children(cursor::CXCursor)
    list = CXCursor[]
    cu_visitor_cb = @cfunction(cu_children_visitor, Cuint, (CXCursor, CXCursor, Ref{Vector{CXCursor}}))
    GC.@preserve list begin; clang_visitChildren(cursor, cu_visitor_cb, pointer_from_objref(list)); end
    return list
end
```

Or even replacing `clang_visitChildren(cursor, cu_visitor_cb, pointer_from_objref(list))` with `ccall((:clang_visitChildren, libclang), UInt32, (CXCursor, CXCursorVisitor, Any), cursor, cu_visitor_cb, list)`

Reply:
> No. `list` can be freed whenever pointer_from_objref(list) returns.

`list` is the return variable of `children`. How could it be GC-collected before returned?

Reply:
> `list` is the return variable of `children`

That doesn't matter. FWIW, even though the compiler should not do this ATM, the compiler can proof in this case that there's no valid use of `pointer_from_objref(list)` before the return therefore there's no use of `list` before the return and so it doesn't need to allocate `list` before the return at all. The return value may also not be used after inlining.

Reply:
> Or even replacing `clang_visitChildren(cursor, cu_visitor_cb, pointer_from_objref(list))` with `ccall((:clang_visitChildren, libclang), UInt32, (CXCursor, CXCursorVisitor, Any), cursor, cu_visitor_cb, list)`

And after this you don't need the `GC.@preserve list` anymore.